feat: implement limit order hook with order tracking and state manage…

datluongductuan opened the pull request:
…ment

- Added a new limit order hook with functionality to track pending fill orders and manage their states.
- Introduced data structures for order information and status.
- Enhanced the BeforeSwap method to process pending limit orders before executing swaps.
- Updated hook registration to include new functionalities and ABI integration.


<!--- Provide a general summary of your changes in the Title above -->

## Why did we need it?
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Release Note
<!--- Please list out special notes (if any) when releasing this commit -->
<!--- Special notes is things like breaking changes and related components, required changes and env configurations in kyber-application, etc. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

datluongductuan committed:
feat: implement limit order hook with order tracking and state management

- Added a new limit order hook with functionality to track pending fill orders and manage their states.
- Introduced data structures for order information and status.
- Enhanced the BeforeSwap method to process pending limit orders before executing swaps.
- Updated hook registration to include new functionalities and ABI integration.

datluongductuan committed:
feat: enhance limit order hook with tick modification and error handling

- Added ModifyTicks method to the Hook interface for pre-swap tick adjustments.
- Implemented logic in the limit order hook to manage liquidity during order fulfillment.
- Introduced ErrorSubLiquidityUnderflow to handle liquidity underflow scenarios.
- Updated NewPoolSimulator to call ModifyTicks for improved pool state management.

datluongductuan committed:
refactor: update order info mapping in limit order hook

- Changed the type of OrderInfos map to use string keys instead of OrderId for better compatibility.
- Updated the logic in the Track method to convert order IDs to string format when populating OrderInfos.
- Removed unnecessary V3Pool reference from BeforeSwapParams in pool simulator.

datluongductuan committed:
refactor: streamline pool simulator cloning logic

- Removed redundant cloning of pool simulator in CalcAmountOut method.
- Updated cloning logic to ensure V3Pool fee adjustments are applied correctly after swap result validation.
- Cleaned up BeforeSwapParams by removing unnecessary PoolID field.

datluongductuan committed:
refactor: correct tick instance checks in limit order hook

- Updated conditionals in ModifyTicks method to properly handle nil checks for lower and upper tick instances.
- Ensured that liquidity adjustments are made correctly when adding liquidity in the limit order hook.

datluongductuan committed:
refactor: clarify liquidity removal note in ModifyTicks method

- Updated comment in ModifyTicks to specify that liquidity is being removed from the pool as the order is going to be filled.
- Removed unused BeforeSwap method to streamline the hook implementation.

datluongductuan committed:
debug

datluongductuan committed:
refactor: clean up query generation in getPoolsListQuery function

- Removed unused import of fmt package.
- Eliminated hardcoded totalValueLockedUSD_gt filter from the GraphQL query.
- Removed debug print statement to streamline the query output.

datluongductuan committed:
debug: add logging for DexID in GetNewPools method

- Introduced a print statement to log the DexID configuration in the GetNewPools method for debugging purposes.

datluongductuan committed:
debug

datluongductuan committed:
fix: ensure hook matches DexID in GetNewPools method

- Added a check to skip processing if the hook's exchange does not match the configured DexID, improving the accuracy of pool updates.

datluongductuan committed:
debug

datluongductuan committed:
debug

datluongductuan committed:
more debug

datluongductuan committed:
debug: update logging in NewBaseHook function

- Modified the print statement in NewBaseHook to log the entire param object instead of just the DexID, enhancing debugging information for the exchange and parameters.

datluongductuan committed:
debug

datluongductuan committed:
refactor: enhance logging and clean up GetNewPools method

- Updated logging in GetNewPools to include hook address for better traceability.
- Removed redundant check for hook and DexID match, simplifying the logic for pool processing.

datluongductuan committed:
refactor: remove debug print statements and update GraphQL query

- Eliminated unnecessary print statements in GetNewPools method for cleaner code.
- Updated the GraphQL query in getPoolsListQuery to include a specific hook address, enhancing the query's precision.

datluongductuan committed:
refactor: change PendingFillOrderLength type to big.Int for improved precision

- Updated the type of PendingFillOrderLength in the Extra struct from uint256.Int to big.Int to enhance numerical handling and precision in order processing.

datluongductuan committed:
refactor: remove debug print statements for cleaner code

- Eliminated print statements in NewBaseHook and getPoolsListQuery functions to enhance code clarity and maintainability.
- Cleaned up logging in GetNewPools method for improved readability.

datluongductuan committed:
test: add limit order pool simulator tests and sample data

- Introduced a new test file for limit order pool simulations, implementing the TestCalcAmountOutWithLoPool function to validate output calculations.
- Added a sample JSON file representing a limit order pool to facilitate testing scenarios.

datluongductuan committed:
refactor: remove hardcoded hook address from GraphQL query

- Removed the hardcoded hook address from the getPoolsListQuery function to improve flexibility and maintainability of the query.

lehainam-dev commented on pkg/liquidity-source/pancake/infinity/cl/pool_simulator.go:
```diff
@@ -49,8 +50,14 @@ func NewPoolSimulator(entityPool entity.Pool, chainID valueobject.ChainID) (*Poo
-	var allowEmptyTicks bool
+	// modify ticks before new pool simulator, some hooks will need this.
+	// In the the original logic, we should do this logic in BeforeSwap in CalcAmountOut, but using here for simplicity.
+	err := hook.ModifyTicks(context.Background(), extra.ExtraTickU256)
```
Seem like we also don't use context inside ModifyTicks. Can we skip this field?

datluongductuan replied:
yeah I removed this ctx from params

datluongductuan committed:
refactor: remove context parameter from ModifyTicks method

- Simplified the ModifyTicks method in the Hook interface and its implementations by removing the context parameter, enhancing code clarity and reducing complexity.

datluongductuan committed:
add debug hook for fast bootstrap
